### ISSUE TITLE
Support IList<ReadOnlyMemory<byte>> SocketAsyncEventArgs

### ISSUE THREAD

**Issue by davidfowl** [dotnet/runtime#49941]
## Background and Motivation

Doing some profiling writing large payloads via Kestrel, about 1.2 % of the overall time is spent converting from `ReadOnlySequence<byte>` to a `List<ArraySegment<byte>>` for the eventual call to the underlying socket API that takes an array of pointers to buffers. Most of the time is spent calling into [`MemoryMarshal.TryGetArray`](https://github.com/dotnet/runtime/blob/82ca681cbac89d813a3ce397e0c665e6c051ed67/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs#L261) to get an `ArraySegment<byte>`  from the `Memory<byte>` most of which is a pointless conversion.

## Proposed API

```diff
namespace System.Net.Sockets
{
    public class SocketAsyncEventArgs
    {
+    public IList<ReadOnlyMemory<byte>>? MemoryList { get; set; }
    }
}
```

## Usage Examples

``` C#
private void SetBufferList(in ReadOnlySequence<byte> buffer)
{
    if (_bufferList == null)
    {
        _bufferList = new List<ReadOnlyMemory<byte>>();
    }

    foreach (ReadOnlyMemory<byte> b in buffer)
    {
        _bufferList.Add(b);
    }

    // The act of setting this list, sets the buffers in the internal buffer list
    BufferList = _bufferList;
}
```

## Risks

More ways of doing the same thing. When to choose MemoryList vs BufferList. Do they overwrite each other? etc.

**Comment by geoffkizer** [comment]
I agree, this would be a good thing to do. Main difficulty is what you highlighted above:

> More ways of doing the same thing. When to choose MemoryList vs BufferList. Do they overwrite each other? etc.

The other consideration here is, we want this to work seamlessly with NetworkStream, so whatever way we choose to expose this in SAEA and other socket APIs will effectively commit us to the same approach in Stream etc. So we need to design this in concert with https://github.com/dotnet/runtime/issues/25344.

Also note, ReadOnlyMemory only covers send cases; presumably we want to support receive here as well. Which complicates things further.

**Comment by geoffkizer** [comment]
All that said: Can we improve the cost of MemoryMarshal.TryGetArray? That seems really high for what should be a very common operation.

**Comment by davidfowl** [comment]
BTW this all came up because I'm looking at this https://github.com/dotnet/aspnetcore/issues/31110

> The other consideration here is, we want this to work seamlessly with NetworkStream, so whatever way we choose to expose this in SAEA and other socket APIs will effectively commit us to the same approach in Stream etc. So we need to design this in concert with #25344.

Are you saying we should use the same types on both? I agree. I wouldn't gate one of the other but they can be designed together.

> Also note, ReadOnlyMemory only covers send cases; presumably we want to support receive here as well. Which complicates things further.

Good point. I think multi-buffer sends are 90% and multi buffer receives are 10% (I pulled those numbers out of thin air btw).

> All that said: Can we improve the cost of MemoryMarshal.TryGetArray? That seems really high for what should be a very common operation.

No idea. @GrabYourPitchforks? The other issue is that Kestrel uses 4K buffers so we end up doing this lots of time doing these on individual 4K chunks.

**Comment by davidfowl** [comment]
BTW there's a bunch of other overhead I hit as well. I filed the issue for the easy one so far.

**Comment by geoffkizer** [comment]
> Are you saying we should use the same types on both? I agree.

Yes. E.g. is it IList or IReadOnlyList? We need to use exactly the same types here across Socket/Stream or we could accidentally cause ourselves a lot of pain.

> I wouldn't gate one of the other but they can be designed together.

Yeah, we could choose to implement this in parts in different ways, we just need to feel confident that the design works in all cases.

**Comment by geoffkizer** [comment]
> BTW there's a bunch of other overhead I hit as well. 

Please file it all :)

**Comment by benaadams** [comment]
Or `ReadOnlySequence<byte>` 🤔

**Comment by davidfowl** [comment]
The problem with ReadOnlySequence\<byte\> is that we'd end up forcing creation of a linked list.

**Comment by benaadams** [comment]
> The problem with ReadOnlySequence<byte> is that we'd end up forcing creation of a linked list.

Not sure why? Change `_bufferListInternal` from `List<ArraySegment<byte>>` to  `List<Memory<byte>>`

https://github.com/dotnet/runtime/blob/1ee59da9f6104c611b137c9d14add04becefdf14/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs#L160-L168

Can then enumerate the `IList<ArraySegment<byte>>` into it and enumerate the `ReadOnlySequence<byte>` into it?

**Comment by davidfowl** [comment]
I meant to create the ReadOnlySequence in the first place.

**Comment by benaadams** [comment]
> I meant to create the ReadOnlySequence in the first place.

True; but might already have one, like if you were forwarding upstream? (like your usage)
```csharp
public async Task InvokeAsync(HttpContext context)
{
    var reader = context.Request.BodyReader;
    var writer = _upstream.Output;
    while (true)
    {
        var readResult = await reader.ReadAsync(default);
        var buffer = readResult.Buffer;
        var isCompleted = readResult.IsCompleted;

        if (buffer.IsEmpty && isCompleted)
        {
            return;
        }

        await writer.WriteAsync(buffer);
    }
```
More mean your usage example could also be an api; and then don't need to create two Lists, since the first thing the `SocketAsyncEventArgs` does is copy the List to another List; so it would save one List creation and add sequence?
```csharp
private void SetBufferList(in ReadOnlySequence<byte> buffer)
{
    if (_bufferListInternal == null)
    {
        _bufferListInternal = new List<ReadOnlyMemory<byte>>();
    }
    else
    {
        _bufferListInternal.Clear();
    }

    foreach (ReadOnlyMemory<byte> b in buffer)
    {
        _bufferListInternal .Add(b);
    }
}
```

**Comment by davidfowl** [comment]
I wouldn't mind adding directly to the buffer list. The setting API is jank and requires me to manage my own list. That would be a decent middleground

**Comment by scalablecory** [comment]
We need a receive variant too, as well as `Send()` and `Receive()` overloads.

@antonfirsov has ideas to make `Send()` and `Receive()` be the first-class APIs and have SAEA eventually wrap them.

**Comment by scalablecory** [comment]
> which mostly defeats the purpose of use passing a pre-pinned memory handle to the networking stack

This is interesting, because I'd expect working with `ArraySegment` to actually be the most efficient here.

What is the overhead of creating a GCHandle to a pre-pinned array? I'd expect it to be close to a no-op. Maybe this is something that can be optimized.

**Comment by scalablecory** [comment]
Whatever types we use here must be coordinated with #25344.

**Comment by davidfowl** [comment]
Allocating a GCHandle isn't a noop, and we shouldn't be doing it. The stack should be re-written on top of `IList<Memory<byte>>` instead of `ArraySegment<byte>`, I'm not sure why that would be contentious (I started messing with it in a local branch).

**Comment by scalablecory** [comment]
> I'm not sure why that would be contentious

I'm not proposing we use `ArraySegment`, I'm just surprised it would be slower than `Memory`.

**Comment by davidfowl** [comment]
The memory we create doesn't have a GC handle since it's using the pinned object heap. This code is allocating a GC handle for the underlying array when it doesn't need to. On top of that, we need to unpack the Memory back into an ArraySegment for each buffer in the linked list chain before setting the result on the SAEA. 

There's all sorts of inefficiencies around multi-buffer sends, this is one of the lower hanging fruit 😄.

**Comment by stephentoub** [comment]
>> What is the overhead of creating a GCHandle to a pre-pinned array? I'd expect it to be close to a no-op. 

> Allocating a GCHandle isn't a noop, and we shouldn't be doing it.

Yeah, it's not free, and it'll be more expensive the more contentious it is.  Creating a GCHandle entails synchronization (best case an interlocked, worst case a lock).  It doesn't really matter what the type of the handle is, nor whether the array is already pinned or not, it's the same appx operation and cost.

| Method |                 type |     Mean |    Error |   StdDev |
|------- |--------------------- |---------:|---------:|---------:|
| Normal |               Normal | 30.88 ns | 0.144 ns | 0.135 ns |
| Normal |               Pinned | 31.36 ns | 0.113 ns | 0.100 ns |
| Normal |                 Weak | 31.53 ns | 0.528 ns | 0.441 ns |
| Normal | WeakT(...)ction [21] | 33.69 ns | 0.151 ns | 0.118 ns |
| Pinned |               Normal | 30.80 ns | 0.071 ns | 0.059 ns |
| Pinned |               Pinned | 31.57 ns | 0.101 ns | 0.090 ns |
| Pinned |                 Weak | 32.15 ns | 0.111 ns | 0.093 ns |
| Pinned | WeakT(...)ction [21] | 33.50 ns | 0.121 ns | 0.107 ns |

```C#
private byte[] _regular = new byte[1024];
private byte[] _pinned = GC.AllocateArray<byte>(1024, pinned: true);

public GCHandleType HandleType { get; set; }

[Benchmark]
[Arguments(GCHandleType.Normal)]
[Arguments(GCHandleType.Pinned)]
[Arguments(GCHandleType.Weak)]
[Arguments(GCHandleType.WeakTrackResurrection)]
public void Normal(GCHandleType type) => GCHandle.Alloc(_regular, type).Free();

[Benchmark]
[Arguments(GCHandleType.Normal)]
[Arguments(GCHandleType.Pinned)]
[Arguments(GCHandleType.Weak)]
[Arguments(GCHandleType.WeakTrackResurrection)]
public void Pinned(GCHandleType type) => GCHandle.Alloc(_pinned, type).Free();
```

**Comment by davidfowl** [comment]
Right, we create the Memory with the [CreateFromPinnedArray](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.createfrompinnedarray?view=net-5.0) to avoid GCHandle overhead and as it turns out, we only avoid it for single buffer writes. 

PS I'm discovering these inefficiencies because I've been looking at this https://github.com/dotnet/aspnetcore/issues/31110 and some other scenarios around reducing memory.

**Comment by antonfirsov** [comment]
> @antonfirsov has ideas to make `SendAsync()` and `Receive()` be the first-class APIs and have SAEA eventually wrap them.

I didn't mean to wrap the `ValueTask` API-s by SAEA API-s. The idea is that the (cancellable) `ValueTask`  variants should be close-to-zero-overhead wrappers around the SAEA stuff, making them the primary choice for every developer who needs async, including Kestrel devs.

I proposed those overloads in https://github.com/dotnet/runtime/issues/48477#issuecomment-784246675, copying them here for better visibility:
```C#
public class Socket
{
    public ValueTask SendAsync(IReadOnlyList<ReadOnlyMemory<byte>> buffers, CancellationToken cancellationToken = default);
    public ValueTask<int> ReceiveAsync(IReadOnlyList<Memory<byte>> buffers, CancellationToken cancellationToken= default);
}
```

@geoffkizer had a valid concern:

> Also note, ReadOnlyMemory only covers send cases; presumably we want to support receive here as well. Which complicates things further.

With the `ValueTask` overloads we can avoid exploding the SAEA's public API further. (We can manage the lists as implementation details of Socket's internal SAEA.)

To provide better experience for callers with a ROS, we may also add a convenience overload for ROS, which wraps the `IList<ReadOnlyMemory<byte>>` variant by populating an internally cached list, which seems to be very easy to implement, and the overhead seems to be acceptable for me:

```C#
    public ValueTask SendAsync(ReadOnlySequence<byte> buffers, CancellationToken cancellationToken = default);
```

**Comment by geoffkizer** [comment]
> With the ValueTask overloads we can avoid exploding the SAEA's public API further.

Yeah, I like this approach a lot.

**Comment by geoffkizer** [comment]
> That's nice but where do IOVectors and WsaBuf structs get stored? You can only stack alloc so many so in our case I'd like to figure out how to reuse a cached set of these structs allocated on the heap once we exceed that limit.

We can stack alloc a decent amount; I think we stack alloc up to 8 today -- @tmds would know for sure. And we could increase this (somewhat) if it makes a difference for real scenarios. 

But regardless, we would still be using a cached SAEA instance under the covers, and that would reuse cached WSABUFs etc. It's just that we wouldn't expose Memory support on SAEA publicly, it would be an internal-only thing.

In other words I don't think there's an issue here.

**Comment by davidfowl** [comment]
> But regardless, we would still be using a cached SAEA instance under the covers, and that would reuse cached WSABUFs etc. It's just that we wouldn't expose Memory support on SAEA publicly, it would be an internal-only thing.

OK then I prefer this approach as well. I thought the goal was to avoid SAEA (which is something I want to do for reads BTW).

**Comment by geoffkizer** [comment]
> I thought the goal was to avoid SAEA (which is something I want to do for reads BTW).

Why is that?

**Comment by geoffkizer** [comment]
BTW, how many buffers would you be using here typically?

I would be careful about approaches that often require using a lot of buffers for scatter/gather -- e.g. >4 to be conservative, or maybe >8. I'm not sure the kernels here do a great job in this case.

**Comment by stephentoub** [comment]
> I thought the goal was to avoid SAEA

SAEA is an implementation detail for the socket methods, a detail we can change over time.  Right now it stores the kitchen sink: if we cache one for use in single-buffer ReadAsync calls, we're still paying for all the state associated with every other operation it can support.  My ideal over time is that we evolve to have some smaller internal data structure on top of which we can build both the socket operations and SAEA, which at that point would be purely legacy / for compat.  Waving my hands.

**Comment by davidfowl** [comment]
> BTW, how many buffers would you be using here typically?
> I would be careful about approaches that often require using a lot of buffers for scatter/gather -- e.g. >4 to be conservative, or maybe >8. I'm not sure the kernels here do a great job in this case.

We use 4K buffers today so 8 buffers would be 32K. Large responses are sometimes more than this so it can get up there in the number of buffers (256 for 1MB). I'm playing around with different buffer sizes and dynamic buffer sizes but that's not likely to be something that we commit anytime soon. Right now, I'm trying to reduce the overhead at all of the layers as much as possible.

> SAEA is an implementation detail for the socket methods, a detail we can change over time. Right now it stores the kitchen sink: if we cache one for use in single-buffer ReadAsync calls, we're still paying for all the state associated with every other operation it can support. My ideal over time is that we evolve to have some smaller internal data structure(s) per operation on top of which we can build both the socket operations and SAEA, which at that point would be purely legacy / for compat. Waving my hands.

Yes, I know this, but we don't use the Task methods, they add overhead we don't need (and no I haven't measured 😄), but I don't see the value in adding overhead especially at the networking layer. We'll kinda do insane things here to avoid it. Right now, for ReadAsync we end up holding our derived `SocketAsyncEventArgs` in the state machine in a pending zero byte read and that's ~350 bytes per connection. I'd like that to be as minimal as possible so I can pool the SAEA for reads as well, instead of having one per connection.

**Comment by geoffkizer** [comment]
> I don't see the value in adding overhead especially at the networking layer.

While it's true that the Task APIs are basically just wrappers over public SAEA today, that will likely change over time, as @stephentoub pointed out. The goal would be to reduce memory usage etc in the Task APIs in a way that is difficult to do via public SAEA. 

In other words, in the future you'll probably want to use Task APIs anyway.

**Comment by davidfowl** [comment]
It's fine to change them but right now the Task APIs don't let us control the scheduling which we care about (actually I'd prefer that directly in the socket APIs themselves). But I'm less concerned about sends than receives. We've already changed the send side to pool SAEA so we have a very small number of those now, but receives mean we need one per connection. 

I'd like to see how we achieve lower memory usage with the Task APIs directly, it'd need to pool something across instances to get the same low footprint.

**Comment by stephentoub** [comment]
> right now the Task APIs don't let us control the scheduling which we care about

What scheduling does SEA support that the Socket APIs don't?

**Comment by tmds** [comment]
On Unix we stackalloc up to 8 for this reason:

https://github.com/dotnet/runtime/blob/400311b032fe5d05b49fb6a813e4a5a60604d8dd/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs#L21-L23

**Comment by davidfowl** [comment]
> What scheduling does SAEA support that the Socket APIs don't?
I misspoke, it isn't the scheduling, it's just the unnecessary overhead in our scenario:

- They try to support concurrent operations and have lots of code to handle those cases (overlapping reads and writes).
- We want to directly schedule our continuation but with the Task based APIs we need to wait for the await to yield before we thread hop again.

**Comment by davidfowl** [comment]
To get this issue back on topic, because I'd like to see a proposal that we all agree with:
- I'd like an API that doesn't regress the pooling we do today.
- I'd like an API that doesn't have hidden allocations in the name of usability.

The problem I have with the SendAsync API is that it will inevitably have to pool WSABufs/IOVectors over a certain size. If those heap allocated arrays are rented from the array pool, I think that would be a big improvement. The next thing would be to pool overlapped objects on windows across different socket implementations (PreallocatedOverlapped). 

I think if we did those, I would feel better about using the SendAsync overloads.

**Comment by geoffkizer** [comment]
I wonder if it would make more sense to just allow you to reuse Socket instances, which would mean you'd reuse the cached SAEAs and PreallocatedOverlapped and other things too.

**Comment by antonfirsov** [comment]
> I wonder if it would make more sense to just allow you to reuse Socket instances, which would mean you'd reuse the cached SAEAs and PreallocatedOverlapped and other things too.

I guess this would mean designing new `Dispose` API-s to which allow avoiding the disposal of cached SAEA-s, so `socket.AcceptAsync(System.Net.Sockets.Socket? acceptSocket)` can reuse them? This looks exposing implementation details, I think extending `SocketAsyncEventArgs` is still a (way much) lesser evil.

**Comment by karelz** [comment]
Triage: We should keep discussing final design for post-6.0. It is reasonable for .NET 7.

**Comment by pepone** [comment]
Would you consider using the same API already used for `QuicStream` 

https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs#L98

This would make it simpler to support TCP and Quick with the same code.

**Comment by davidfowl** [comment]
Now that we have in-line arrays I wonder if these should be span based APIs.

**Comment by stephentoub** [comment]
> Now that we have in-line arrays I wonder if these should be span based APIs.

Which specifically? Async and span don't mix well.

**Comment by davidfowl** [comment]
It wouldn’t work for SocketAsyncEventArgs but IIRC the underlying methods don’t need to pin the array passed in, They are usually copied. 

I know the lowest API we have is SAEA so this approach doesn’t flow through Socket.SendAsync in a clean way…

**Comment by stephentoub** [comment]
> It wouldn’t work for SocketAsyncEventArgs but IIRC the underlying methods don’t need to pin the array passed in, They are usually copied.
> 
> I know the lowest API we have is SAEA so this approach doesn’t flow through Socket.SendAsync in a clean way…

I'm still not understanding. Can you sketch what it is you're envisioning? If the operation can't complete immediately and fully, such that the operation needs to pend and try again later, the relevant data or buffers still needs to be available.

**Comment by davidfowl** [comment]
I was thinking about how we end up calling these APIs:
- https://github.com/dotnet/runtime/blob/bd83e17052d3c09022bad1d91dca860ca6b27ab9/src/libraries/Common/src/Interop/Windows/WinSock/Interop.WSASend.cs#L24C8-L24C8
- https://github.com/dotnet/runtime/blob/bd83e17052d3c09022bad1d91dca860ca6b27ab9/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs#L284

> I'm still not understanding. Can you sketch what it is you're envisioning? If the operation can't complete immediately and fully, such that the operation needs to pend and try again later, the relevant data or buffers still needs to be available.

The buffers themselves are pinned for the lifetime of the async operation but the holder array just needs to be pinned for the length of the synchronous call.

> Do you just mean an API like SendAsync(ReadOnlySpan<ReadOnlyMemory<byte>> and it would copy that span to its own possibly-pooled heap object?

Yes exactly. We would flow this span all the way down to the OS API calls then translate it into the relevant Span<WSABuffer/IOVector>. 

This doesn't gel with the current SAEA BufferList property, and I haven't thought through how to flow it from the user code down to the OS API.

Contrived example:

```c#
using System.Net;
using System.Net.Sockets;
using System.Runtime.CompilerServices;

var socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
socket.Bind(new IPEndPoint(IPAddress.Loopback, 5000));
socket.Listen();

var client = socket.Accept();

var buffers = new Buffers();
buffers[0] = GetHeaders();
buffers[1] = GetBody();

await client.SendAsync(buffers);

ReadOnlyMemory<byte> GetHeaders() => default;
ReadOnlyMemory<byte> GetBody() => default;

[InlineArray(2)]
struct Buffers
{
    private ReadOnlyMemory<byte> _buffers;
}

static class Extensions
{
    public static ValueTask<int> SendAsync(this Socket socket, ReadOnlySpan<ReadOnlyMemory<byte>> buffers) => default;
}
```

**Comment by stephentoub** [comment]
>> Do you just mean an API like SendAsync(ReadOnlySpan<ReadOnlyMemory> and it would copy that span to its own possibly-pooled heap object?

> Yes exactly. We would flow this span all the way down to the OS API calls then translate it into the relevant Span<WSABuffer/IOVector>.

Just to be extra clear, though, it's not just about pinning. The list of buffers needs to be available asynchronously from the current stack if the operation doesn't complete synchronously, which means copying it somewhere.

**Comment by davidfowl** [comment]
> Just to be extra clear, though, it's not just about pinning. The list of buffers needs to be available asynchronously from the current stack if the operation doesn't complete synchronously, which means copying it somewhere.

Are you talking about the Linux case where we queue the operation? Or do you mean in both cases?

**Comment by stephentoub** [comment]
> Are you talking about the Linux case where we queue the operation?

I'm talking about the non-Windows case where if the native call on the non-blocking socket returns EAGAIN / EWOULDBLOCK, we wait on an epoll / kqueue and try again when notified that we can make forward progress with the operation.

**Comment by davidfowl** [comment]
I see, yes that makes sense. It would force us to copy the array in that case.
